### ISSUE TITLE
Make tps-listener more resilient to issues in doppler

### ISSUE THREAD

**Issue by keymon** [cloudfoundry-incubator/tps#5]
Issue description
--------------------

We had some issues with the tps-listener when [is calling doppler to get some metrics here])https://github.com/cloudfoundry-incubator/tps/blob/eadeccb88608257fad1c154d1623beb9ce1571ca/handler/lrpstats/lrpstats.go#L73). 

These metrics from doppler seem to be something  optional for tps-listener, as it simply carries on if the request fails, but it is implemented **with a synchronous blocking call without timeout**. If for some reason the doppler is unresponsive, misconfigured, the network is dropping packages or the request is really slow responding, `tps-listener` will lock the request until the other side closes  it or it timeouts. 

The `tps-listener` is called by CF CloudController when the user executes `cf app <app>` or the application is deployed. The CF Cloud Controller would timeout. In our case, this meant that any deployment was failing with the following error (also `cf app <appname>`) would fail:

```
Could not fetch instance count: Server error, status code: 503, error code: 200002, message: Stats unavailable: Stats server temporarily unavailable.
```

Expected behaviour
-------------------------

The request to doppler should timeout quick. Circuit breaking logic would be a plus.

How to reproduce
----------------------

You can start tps-listener pointing to a doppler server that drops the connection (e.g. www.google.com:6666)

``` 
/var/vcap/packages/tps/bin/tps-listener \
    -bbsClientCert=/var/vcap/jobs/tps/config/certs/bbs/client.crt \
    -bbsClientKey=/var/vcap/jobs/tps/config/certs/bbs/client.key \
    -bbsCACert=/var/vcap/jobs/tps/config/certs/bbs/ca.crt \
    -bbsAddress=https://bbs.service.cf.internal:8889 \
    -consulCluster=http://127.0.0.1:8500 \
    -dropsondePort=3457 \
    -maxInFlightRequests=200 \
    -listenAddr=0.0.0.0:6666 \
    -debugAddr=0.0.0.0:6667 \
    -logLevel=info \
    -skipSSLVerification=true \
    -trafficControllerURL=wss://google.com:6666 
```

Stack journey
----------------

After deployment or running  `cf app <appname>` the client executes this query:

``` 
cf curl /v2/apps/3bf24a38-cfdb-4e55-ad38-83e1ad7823f7/stats
``` 

When using diego, CC will query tps-listener with:

```
curl -m 5 -I http://10.0.16.54:1518/v1/actual_lrps/3bf24a38-cfdb-4e55-ad38-83e1ad7823f7-400ad2ef-1056-4296-806e-a851df740dc5/stats -H 'Authorization: bearer ....'
```

Which will try to contact doppler. This query must return quick (fail or not) or it would lock tps.:

``` 
curl -m 5 -I \
    https://doppler.example.com:443/apps/3bf24a38-cfdb-4e55-ad38-83e1ad7823f7/containermetrics \
    -H 'Authorization: bearer ...'
```

**Comment by ematpl** [comment]
Thanks for the report, @keymon! That behavior does sound frustrating for developers on the platform.

As @jfmyers9 mentioned in the Diego slack channel, we're using the [`Consumer` from the `noaa` library](https://github.com/cloudfoundry/noaa/blob/master/consumer/consumer.go), which constructs its own `http.Client` instance and therefore constrains our ability as a client to control the timeout behavior at the HTTP layer. Depending on the behavior that would be most useful, we might be able to persuade @Jim-Campbell and @wfernandes to add an additional constructor function for the `Consumer` that allows the appropriate timeout configuration, or even the injection of the entire `http.Client`.

I think it would still be possible for the TPS-Listener to build an effective client-side circuit breaker around that limitation, but I'd like @SocalNick and @utako from the CAPI team to weigh in both as the new owners of the TPS codebase after its recent transfer to [capi-release](https://github.com/cloudfoundry/capi-release) and as the owners of the CC API to decide how 'optional' the container metrics are in the app stats response. If there's a good way for TPS and then CC to convey to clients that the container metrics are missing because the doppler circuit is open, that could be an acceptable solution even without changes to the `noaa` library.

Thanks again,
Eric, CF Runtime Diego PM

**Comment by wfernandes** [comment]
@ematpl We would be glad to assist in this if it helps solve some pain. Adding a constructor to pass in a pre-built `http.Client` seems straightforward.

**Comment by Jim-Campbell** [comment]
+1 I wrote a story. <https://www.pivotaltracker.com/story/show/119835033>

*and*

I've always been curious about this pattern. After a Doppler outage took
down 'cf push' in prod, I wondered about the front end of app monitoring
reaching all the way around to the backed of metric output to do app
monitoring. Is there a 'ELI5' description about why this is the way to go?

Jim


On Wed, May 18, 2016 at 9:35 AM, Warren Fernandes <notifications@github.com>
wrote:

> @ematpl <https://github.com/ematpl> We would be glad to assist in this if
> it helps solve some pain. Adding a constructor to pass in a pre-built
> http.Client seems straightforward.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cloudfoundry-incubator/tps/issues/5#issuecomment-220065795>
>



-- 
Jim Campbell | Product Manager | Cloud Foundry | Pivotal.io | 303.618.0963

**Comment by ematpl** [comment]
Thanks, @wfernandes! What interface would you be considering for that new constructor? Given how much internal wiring the current Consumer constructor does with its existing proxy and TLS arguments, I'd be concerned that allowing the caller also to specify an `*http.Client` could lead to skewed configuration between the HTTP calls and the websocket calls. But you also can't necessarily pluck the proxy and TLS config off of the `Client`, since those are configured on an `http.Transport`, and technically that's just a `RoundTripper` interface on the `Client`.

**Comment by ematpl** [comment]
@Jim-Campbell Since the Diego cells are continually reporting container metrics to the loggregator system, which has an API for clients to retrieve them, we regard the loggregator system as the canonical source for clients to retrieve those container metrics. Then Diego can delegate that responsibility to that system and avoid duplicating an API (which can then drift and otherwise become inconsistent with the loggregator one).

From the perspective of CC, those metrics are only ever retrieved when a client calls `/v2/apps/:guid/stats`, and other calls that report on app instance health, such as `/v2/apps/:guid/instances`, rely only on the instance health information available directly from the Diego BBS API. The CLI calls that `stats` endpoint frequently at the end of `cf push`, though, so if it fails it can cause the `cf` command to fail. Maybe that's an argument that the CLI should instead also rely on the container-metrics endpoint on the loggregator traffic-controller for those container metrics instead of calling the CC's stats endpoint, but in any case it should likely degrade more gracefully in the case of errors from the backend.

**Comment by utako** [comment]
Hi @keymon,

We have some stories coming up to bring a lot of the CC bridge components into the Cloud Controller codebase. We hope to incept on it sometime in the coming month. We hope to sync up with @Jim-Campbell, the Loggregator team, @ematpl, the Diego, about the current state of Doppler/NOAA and how to keep this in mind when we're doing that chunk of work.

Thanks for submitting this issue.

@rizwanreza and @utako, CF CAPI Team

**Comment by Jim-Campbell** [comment]
I'll be in SF 6/22 and 6/23. Why don't you schedule some time on my
calendar and we can connect about this.

On Fri, Jun 10, 2016 at 12:15 PM, Utako Ueda <notifications@github.com>
wrote:

> Hi @keymon <https://github.com/keymon>,
>
> We have some stories coming up to bring a lot of the CC bridge components
> into the Cloud Controller codebase. We hope to incept on it sometime in the
> coming month. We hope to sync up with @Jim-Campbell
> <https://github.com/Jim-Campbell>, the Loggregator team, @ematpl
> <https://github.com/ematpl>, the Diego, about the current state of
> Doppler/NOAA and how to keep this in mind when we're doing that chunk of
> work.
>
> Thanks for submitting this issue.
>
> @rizwanreza <https://github.com/rizwanreza> and @utako
> <https://github.com/utako>, CF CAPI Team
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudfoundry-incubator/tps/issues/5#issuecomment-225256516>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ANIc8P_L2AdL1P4IQu4Hpc5gxHt95JJfks5qKam7gaJpZM4IgSo8>
> .
>



-- 
Jim Campbell | Product Manager | Cloud Foundry | Pivotal.io | 303.618.0963

**Comment by Jim-Campbell** [comment]
Story to solve this is now [this one in the Loggregator backlog.]https://www.pivotaltracker.com/story/show/120317073). We'll close this issue when we Accept that one. Recommend that the [original story](https://www.pivotaltracker.com/story/show/119835033) be closed or deleted.

**Comment by SocalNick** [comment]
Closing in favor of https://www.pivotaltracker.com/n/projects/993188/stories/120317073

**Comment by vvraskin** [comment]
Hi @SocalNick 

could you please point to the commit that fixes this issue?
The stories that are mentioned here dont seem to be public. 

Thanks.

**Comment by jszroberto** [comment]
@vvraskin Apparently, this is the commit which fixed the issue: https://github.com/cloudfoundry/noaa/commit/8d52d64e4b76ab53ea11117dbe109201a025e6ac
